From 44167fc51e0f743c86627be8e9ff6db75aa192cd Mon Sep 17 00:00:00 2001
From: stone.liu <stone.liu@dniserver.dnish.net>
Date: Fri, 28 Oct 2011 10:40:54 -0400
Subject: [PATCH 111/137] netfilter: supporting established UDP NAT Outbound refresh feature

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
REQ-6: The NAT mapping Refresh Direction MUST have a "NAT Outbound refresh
behavior" of "True".

NETGEAR Requirements (UDP/TCP) ---
        1. NAT Outbound Refresh Behavior: MUST;
        2. Inbound Refresh Behavior: MUST NOT.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

DNI Common Module Patch Review Checklist
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here are some basic things that I (Jonathan Chang) have checked
in order to ensure the quality of common modules.

1: The syntax and comment are compliant.

2: No memory leak.
   * memory allocated by malloc() will be freed
   * memory allocated by calloc() will be freed

3: No deadlock such as
   * No function call deadlock.
   * No semaphore deadlock.
   * No tag or label deadlock.

4: No stack overflow caused by
   * infinite recursion
   * too large local array variables

5: coding completeness, such as
   * conditions completeness
   * functions completeness
   * update completeness
   * memory resources completeness
   * sessions resources completeness
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Signed-off-by: Haiyue.Wang <haiyue@dniserver.dnish.net>
Signed-off-by: Jack Lin <jack.lin@dniserver.dnish.net>
Signed-off-by: Jonathan Chang <jonathan@dniserver.dnish.net>

Conflicts:

	net/netfilter/nf_conntrack_proto_udp.c

Signed-off-by: stone.liu <stone.liu@dniserver.dnish.net>
---
 net/netfilter/nf_conntrack_proto_udp.c |   26 ++++++++++++++++++++++++++
 1 files changed, 26 insertions(+), 0 deletions(-)

diff --git a/net/netfilter/nf_conntrack_proto_udp.c b/net/netfilter/nf_conntrack_proto_udp.c
index ba24655..2a3af10 100644
--- a/net/netfilter/nf_conntrack_proto_udp.c
+++ b/net/netfilter/nf_conntrack_proto_udp.c
@@ -85,6 +85,32 @@ static int udp_packet(struct nf_conn *ct,
 #if defined(CONFIG_RTL_NF_CONNTRACK_GARBAGE_NEW)
 		nf_ct_refresh_acct_udp(ct, ctinfo, skb, nf_ct_udp_timeout_stream, "ASSURED"); // assured
 #else
+		/*
+		 * REQ-6: The NAT mapping Refresh Direction MUST have a "NAT Outbound refresh
+		 * behavior" of "True".
+		 *
+		 * NETGEAR Requirements (UDP/TCP) --- 1. NAT Outbound Refresh Behavior: MUST;
+		 * 2. Inbound Refresh Behavior: MUST NOT.
+		 *
+		 * Do some optimization from `CTINFO2DIR':
+		 * #define CTINFO2DIR(ctinfo)   \
+		 *     ((ctinfo) >= IP_CT_IS_REPLY ? IP_CT_DIR_REPLY : IP_CT_DIR_ORIGINAL)
+		 */
+		if (sysctl_conntrack_refresh_outbound_only != 0) {
+			if (ct->status & IPS_SRC_NAT) {
+				if (dir == IP_CT_DIR_REPLY)
+					refresh = 0;
+				else
+					refresh = 1;
+			} else if (ct->status & IPS_DST_NAT) {
+				if (dir == IP_CT_DIR_ORIGINAL)
+					refresh = 0;
+				else
+					refresh = 1;
+			} else
+				refresh = 1;
+		}
+
 		if (refresh)
 		nf_ct_refresh_acct(ct, ctinfo, skb, nf_ct_udp_timeout_stream);
 #endif			
-- 
1.7.4.1

